### PR TITLE
Fix variant without options values generation

### LINKED ISSUE #9576: Variant generation throws exception when option has no values
**Sylius version affected**: 1.2.2

**Exception Message**  

> The set builder requires a single array of one or more array sets.

**Steps to reproduce**  

1. Create an option (DO NOT ADD ANY VALUES).
2. Create configurable product (select option created in previous step). 
3. Press Manage variants -> generate on created product.
4. Boom exception.

**Possible Solution**  
Require at least one value when creating options.

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | 1.3
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes #9576
| License         | MIT

Screenshot 2019-03-22 at 09 01 05

### CODE REVIEW

File: features/product/managing_product_variants/generating_product_variant_without_value.feature [hunk 1]
```diff
@@ -0,0 +1,17 @@
+@managing_product_variants
+    Feature: Generating product variants without any values
```
Comment (Zales0123) [inline features/product/managing_product_variants/generating_product_variant_without_value.feature]: Wrong indent here and in the whole file

File: features/product/managing_product_variants/generating_product_variant_without_value.feature [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        In order to sell variants without any values of a single product
```
Comment (Zales0123) [inline features/product/managing_product_variants/generating_product_variant_without_value.feature]: I think it should be something like `In order to not be able to generate variants without any options values available`

File: features/product/managing_product_variants/generating_product_variant_without_value.feature [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+        As an Administrator
+        I want to generate variants without any values
```
Comment (Zales0123) [inline features/product/managing_product_variants/generating_product_variant_without_value.feature]: `I want to not be able to generate variants without any option values`

File: src/Sylius/Behat/Context/Setup/ProductContext.php [hunk 4]
```diff
@@ -931,4 +951,6 @@ private function createChannelPricingForChannel(int $price, ChannelInterface $ch
 
         return $channelPricing;
     }
+
+
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Setup/ProductContext.php]: Redundant

File: src/Sylius/Behat/Context/Ui/Admin/BrowsingProductVariantsContext.php [hunk 5]
```diff
@@ -74,6 +74,8 @@ public function theProductVariantShouldNotAppearInTheShop($productVariantCode, P
 
     /**
      * @Then the :product product should have no variants
+     * @Given (this product) should have no variants
+
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/BrowsingProductVariantsContext.php]: Redundant

File: src/Sylius/Behat/Context/Ui/Admin/BrowsingProductVariantsContext.php [hunk 6]
```diff
@@ -74,6 +74,8 @@ public function theProductVariantShouldNotAppearInTheShop($productVariantCode, P
 
     /**
      * @Then the :product product should have no variants
+     * @Given (this product) should have no variants
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/BrowsingProductVariantsContext.php]: `@Given` -> `@Then`

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 7]
```diff
@@ -378,6 +378,17 @@ public function iShouldBeNotifiedThatPricesInAllChannelsMustBeDefined()
...
 
+    /**
+     * @Then I should be notified that it has been impossible generate
+     */
+    public function iShouldBeNotifiedThatItHasBeenImpossibleGenerate(): void
+    {
+        /** @var CreatePageInterface|UpdatePageInterface $currentPage */
+        $currentPage = $this->currentPageResolver->getCurrentPageWithForm([$this->createPage, $this->updatePage]);
+
+        Assert::same($currentPage->getValidationMessageForForm(), 'Variant without options can not exist');
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: `can not` -> `cannot`?

File: src/Sylius/Behat/Context/Setup/ProductContext.php [hunk 8]
```diff
@@ -533,6 +533,26 @@ public function thisProductHasOptionWithValues(ProductInterface $product, $optio
         $this->objectManager->flush();
     }
 
+    /**
+     * @Given /^(this product) has option "([^"]+)" without values$/
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Setup/ProductContext.php]: I wonder, what if you put this definition in the previous function `thisProductHasOptionWithValues`? It should then map `...$values` to an empty array and we can get rid of this duplicated code 🎉

**Comment by Zales0123** [comment]
The base of this pull-request was changed, you need fetch and reset your local branch 
if you want to add new commits to this pull request. **Reset before you pull, else commits
may become messed-up.**

Unless you added new commits (to this branch) locally that you did not push yet, 
execute `git fetch origin && git reset "without-any-values"` to update your local branch.
 
Feel free to ask for assistance when you get stuck :+1:

File: src/Sylius/Behat/Context/Ui/Admin/BrowsingProductVariantsContext.php (hunk 6 above)
Comment (GSadee) [reply src/Sylius/Behat/Context/Ui/Admin/BrowsingProductVariantsContext.php]: `(this product) should have no variants` -> `/^(this product) should have no variants$/`

File: src/Sylius/Behat/Context/Ui/Admin/BrowsingProductVariantsContext.php (hunk 5 above)
Comment (lchrusciel) [reply src/Sylius/Behat/Context/Ui/Admin/BrowsingProductVariantsContext.php]: the blank line is still redundant :) Please, remove it.

File: src/Sylius/Behat/Context/Setup/ProductContext.php (hunk 4 above)
Comment (lchrusciel) [reply src/Sylius/Behat/Context/Setup/ProductContext.php]: Both lines empty lines above are still redundant. Please, remove it.

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 9]
```diff
@@ -378,6 +378,14 @@ public function iShouldBeNotifiedThatPricesInAllChannelsMustBeDefined()
         );
     }
 
+    /**
+     * @Then I should be notified that it has been impossible generate
+     */
+    public function iShouldBeNotifiedThatItHasBeenImpossibleGenerate(): void
+    {
+        $this->notificationChecker->checkNotification('Cannot generate variants for a product without options', NotificationType::failure());
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: `for a product without options` -> `for a product without options values`

File: src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/GenerateProductVariantsSubscriber.php [hunk 10]
```diff
@@ -47,6 +52,10 @@ public function preSetData(FormEvent $event): void
 
         Assert::isInstanceOf($product, ProductInterface::class);
 
-        $this->generator->generate($product);
+        try{
+            $this->generator->generate($product);
+        } catch (\InvalidArgumentException $exception) {
+            $this->session->getFlashBag()->add('error',$exception->getMessage());
```
Comment (Zales0123) [inline src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/GenerateProductVariantsSubscriber.php]: Missing space after coma

File: src/Sylius/Component/Product/Generator/ProductVariantGenerator.php [hunk 11]
```diff
@@ -56,6 +56,10 @@ public function generate(ProductInterface $product): void
             }
         }
 
+        if (empty($optionSet)) {
+            throw new \InvalidArgumentException('Cannot generate variants for a product without options');
```
Comment (Zales0123) [inline src/Sylius/Component/Product/Generator/ProductVariantGenerator.php]: `\InvalidArgumentException` is ok right now, but we can think about something more sophisticated, like `VariantWithNoOptionsValuesException`

File: features/product/managing_product_variants/generating_product_variant_without_value.feature [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a "Wyborowa Vodka" configurable product
+        And this product has option "Taste" without values
+        And I am logged in as an administrator
+
+    @ui
+    Scenario: Generating a product variant for product without options values
+        When I want to generate new variants for this product
+        Then I should be notified that it has been impossible generate
```
Comment (Zales0123) [inline features/product/managing_product_variants/generating_product_variant_without_value.feature]: I'm missing a step "And I should not be able to generate any variants" that would check if the **Generate** button is disabled - it should be for sure 😄

Comment (Tomanhez) [reply features/product/managing_product_variants/generating_product_variant_without_value.feature]: Is it possible to spec button?

Comment (Zales0123) [reply features/product/managing_product_variants/generating_product_variant_without_value.feature]: It's possible to check in Behat is it disable or not 😄

File: features/product/managing_product_variants/generating_product_variant_without_value.feature [hunk 13]
```diff
@@ -0,0 +1,16 @@
+@managing_product_variants
+Feature: Generating product variants without any values
+    In order to not be able to generate variants without any options values available
+    As an Administrator
+    I want to not be able to generate variants without any option values
```
Comment (CoderMaggie) [inline features/product/managing_product_variants/generating_product_variant_without_value.feature]: That's lovely 😂

File: features/product/managing_product_variants/generating_product_variant_without_value.feature [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a "Wyborowa Vodka" configurable product
+        And this product has option "Taste" without values
```
Comment (CoderMaggie) [inline features/product/managing_product_variants/generating_product_variant_without_value.feature]:
```suggestion
        And this product has an option "Taste" without any values
```

File: features/product/managing_product_variants/generating_product_variant_without_value.feature [hunk 15]
```diff
@@ -0,0 +1,16 @@
+@managing_product_variants
+Feature: Generating product variants without any values
```
Comment (CoderMaggie) [inline features/product/managing_product_variants/generating_product_variant_without_value.feature]: It's not the feature you get.
```suggestion
Feature: Preventing the generation of product variants from options without any values
```

File: features/product/managing_product_variants/preventing_the_generation_of_products_variants_from_options_without_any_values.feature [hunk 16]
```diff
@@ -0,0 +1,17 @@
+@managing_product_variants
+Feature:Preventing the generation of product variants from options without any values
+    In order to not be able to generate variants without any options values available
```
Comment (Zales0123) [inline features/product/managing_product_variants/preventing_the_generation_of_products_variants_from_options_without_any_values.feature]: Still too similar to the last line :) I would go with `In order to prevented from invalid variant generation`

File: features/product/managing_product_variants/preventing_the_generation_of_products_variants_from_options_without_any_values.feature [hunk 17]
```diff
@@ -0,0 +1,17 @@
+@managing_product_variants
+Feature:Preventing the generation of product variants from options without any values
```
Comment (Zales0123) [inline features/product/managing_product_variants/preventing_the_generation_of_products_variants_from_options_without_any_values.feature]: Missing space after `Feature:`

File: features/product/managing_product_variants/preventing_the_generation_of_products_variants_from_options_without_any_values.feature [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+    I want to not be able to generate variants without any option values
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a "Wyborowa Vodka" configurable product
+        And this product has an option "Taste" without any values
+        And I am logged in as an administrator
+
+    @ui
+    Scenario: Generating a product variant for product without options values
```
Comment (Zales0123) [inline features/product/managing_product_variants/preventing_the_generation_of_products_variants_from_options_without_any_values.feature]: `Trying to generate a product variant for a product without options values`

File: src/Sylius/Behat/Context/Setup/ProductContext.php [hunk 19]
```diff
@@ -923,4 +914,26 @@ private function createChannelPricingForChannel(int $price, ChannelInterface $ch
 
         return $channelPricing;
     }
+
+    private function addOptionAndVariantIfExist(ProductInterface $product, string $optionName, array $values): void
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Setup/ProductContext.php]: `addOptionToProduct` is sufficient, I suppose

File: features/product/managing_product_variants/preventing_the_generation_of_products_variants_from_options_without_any_values.feature [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+        When I want to generate new variants for this product
+        Then I should be notified that it has been impossible generate
```
Comment (Zales0123) [inline features/product/managing_product_variants/preventing_the_generation_of_products_variants_from_options_without_any_values.feature]: `it has been impossible generate` -> `it has been impossible to generate`

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 21]
```diff
@@ -378,6 +378,14 @@ public function iShouldBeNotifiedThatPricesInAllChannelsMustBeDefined()
         );
     }
 
+    /**
+     * @Then I should be notified that it has been impossible generate
+     */
+    public function iShouldBeNotifiedThatItHasBeenImpossibleGenerate(): void
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: `iShouldBeNotifiedThatItHasBeenImpossibleToGenerate`

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 22]
```diff
@@ -465,6 +473,14 @@ public function iShouldBeNotifiedThatItHasBeenSuccessfullyGenerated()
         $this->notificationChecker->checkNotification('Success Product variants have been successfully generated.', NotificationType::success());
     }
 
+    /**
+     * @Given /^I should not be able to generate any variants$/
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: `@Then I should not be able to generate any variants`

File: src/Sylius/Behat/Page/Admin/ProductVariant/GeneratePageInterface.php [hunk 23]
```diff
@@ -28,4 +28,6 @@ public function removeVariant(int $nth): void;
     public function getValidationMessage(string $element, int $position): string;
 
     public function getPricesValidationMessage(int $position): string;
+
+    public function isPossibleToGenerate(): bool;
```
Comment (Zales0123) [inline src/Sylius/Behat/Page/Admin/ProductVariant/GeneratePageInterface.php]: `isGenerationPossible`?

File: src/Sylius/Behat/Page/Admin/ProductVariant/GeneratePage.php [hunk 24]
```diff
@@ -82,4 +82,9 @@ private function getValidatedField(NodeElement $element): NodeElement
 
         return $element;
     }
+
+    public function isPossibleToGenerate(): bool
+    {
+        return ($this->getDocument()->find('css','button:contains("Generate")')->getAttribute('disabled') !== 'disabled') ? true : false;
```
Comment (Zales0123) [inline src/Sylius/Behat/Page/Admin/ProductVariant/GeneratePage.php]: To make it shorter and cleaner: 

```php
$generateButton = $this->getDocument()->find('css','button:contains("Generate")');

return !$generateButton->hasAttribute('disabled');
```

File: src/Sylius/Bundle/AdminBundle/Resources/views/ProductVariant/generate.html.twig [hunk 25]
```diff
@@ -24,11 +24,10 @@
     {% include '@SyliusAdmin/ProductVariant/Generate/_form.html.twig' %}
     <div class="ui basic segment">
         <div class="ui buttons">
-            <button class="ui labeled icon primary button" type="submit"><i class="random icon"></i>{{- 'sylius.ui.generate'|trans -}}</button>
+            <button class="ui labeled icon primary button" type="submit" {% if form.variants.children is empty %} disabled {% endif %}><i class="random icon"></i>{{- 'sylius.ui.generate'|trans -}}</button>
```
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/views/ProductVariant/generate.html.twig]: Remove spaces around `disabled` to make a result HTML nicer

File: src/Sylius/Component/Product/Generator/ProductVariantGenerator.php [hunk 26]
```diff
@@ -56,6 +56,10 @@ public function generate(ProductInterface $product): void
             }
         }
 
+        if (empty($optionSet)) {
+            throw new \InvalidArgumentException('Cannot generate variants for a product without options values');
```
Comment (Zales0123) [inline src/Sylius/Component/Product/Generator/ProductVariantGenerator.php]: I think it should be a translation key

File: features/product/managing_product_variants/preventing_the_generation_of_products_variants_from_options_without_any_values.feature (hunk 16 above)
Comment (CoderMaggie) [reply features/product/managing_product_variants/preventing_the_generation_of_products_variants_from_options_without_any_values.feature]: The other way round, the last sentence (I want to) should be changed.

File: features/product/managing_product_variants/preventing_the_generation_of_products_variants_from_options_without_any_values.feature [hunk 27]
```diff
@@ -0,0 +1,17 @@
...
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a "Wyborowa Vodka" configurable product
+        And this product has an option "Taste" without any values
+        And I am logged in as an administrator
+
+    @ui
+    Scenario: Trying to generate a product variant for a product without options values
+        When I want to generate new variants for this product
+        Then It has been impossible to generate
```
Comment (Zales0123) [inline features/product/managing_product_variants/preventing_the_generation_of_products_variants_from_options_without_any_values.feature]: `Then I should be notified that it has been impossible to generate variants`

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 28]
```diff
@@ -465,6 +473,14 @@ public function iShouldBeNotifiedThatItHasBeenSuccessfullyGenerated()
         $this->notificationChecker->checkNotification('Success Product variants have been successfully generated.', NotificationType::success());
     }
 
+    /**
+     * @Then /^I should not be able to generate any variants$/
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: Regexp is not needed

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 29]
```diff
@@ -465,6 +473,14 @@ public function iShouldBeNotifiedThatItHasBeenSuccessfullyGenerated()
...
+     */
+    public function iShouldNotBeAbleToGenerateAnyVariantsToThisProduct()
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: Missing `: void`

Comment (Zales0123) [reply src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: `iShouldNotBeAbleToGenerateAnyVariants`

File: src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/GenerateProductVariantsSubscriber.php [hunk 30]
```diff
@@ -47,6 +53,10 @@ public function preSetData(FormEvent $event): void
 
         Assert::isInstanceOf($product, ProductInterface::class);
 
-        $this->generator->generate($product);
+        try{
```
Comment (Zales0123) [inline src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/GenerateProductVariantsSubscriber.php]: Missing space after `try`

File: src/Sylius/Bundle/ProductBundle/Resources/config/services/form.xml [hunk 31]
```diff
@@ -205,6 +205,7 @@
 
         <service id="sylius.form.event_subscriber.product_variant_generator" class="Sylius\Bundle\ProductBundle\Form\EventSubscriber\GenerateProductVariantsSubscriber">
             <argument type="service" id="sylius.generator.product_variant" />
+            <argument type="service" id="session"/>
```
Comment (Zales0123) [inline src/Sylius/Bundle/ProductBundle/Resources/config/services/form.xml]: Missing space before `/>`

File: src/Sylius/Component/Resource/Exception/VariantWithNoOptionsValuesException.php [hunk 32]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Resource\Exception;
+
+class VariantWithNoOptionsValuesException extends \Exception
```
Comment (Zales0123) [inline src/Sylius/Component/Resource/Exception/VariantWithNoOptionsValuesException.php]: Missing `final`

File: src/Sylius/Component/Resource/Exception/VariantWithNoOptionsValuesException.php [hunk 33]
```diff
@@ -0,0 +1,14 @@
...
+{
+
```
Comment (Zales0123) [inline src/Sylius/Component/Resource/Exception/VariantWithNoOptionsValuesException.php]: Extra blank line

File: features/product/managing_product_variants/preventing_the_generation_of_products_variants_from_options_without_any_values.feature [hunk 34]
```diff
@@ -0,0 +1,17 @@
...
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a "Wyborowa Vodka" configurable product
+        And this product has an option "Taste" without any values
+        And I am logged in as an administrator
+
+    @ui
+    Scenario: Trying to generate a product variant for a product without options values
+        When I want to generate new variants for this product
+        Then I should be notified that it has been impossible to generate variants
```
Comment (CoderMaggie) [inline features/product/managing_product_variants/preventing_the_generation_of_products_variants_from_options_without_any_values.feature]:
```suggestion
        Then I should be notified that variants cannot be generated from options without any values
```

File: src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/GenerateProductVariantsSubscriber.php (hunk 30 above)
Comment (Zales0123) [reply src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/GenerateProductVariantsSubscriber.php]: Still relevant

File: src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/GenerateProductVariantsSubscriber.php [hunk 35]
```diff
@@ -47,6 +53,11 @@ public function preSetData(FormEvent $event): void
 
         Assert::isInstanceOf($product, ProductInterface::class);
 
-        $this->generator->generate($product);
+        try {
+            $this->generator->generate($product);
+
```
Comment (Zales0123) [inline src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/GenerateProductVariantsSubscriber.php]: Extra blank line 💃

**Comment by Zales0123** [comment]
Thank you, Tomasz! :1st_place_medal:

**Pull request merged**